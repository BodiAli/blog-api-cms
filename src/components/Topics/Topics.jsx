import { useState, Fragment } from "react";

export default function Topics() {
  const [topicValue, setTopicValue] = useState("");
  const [topics, setTopics] = useState([]);

  return (
    <>
      <form
        method="post"
        onSubmit={(e) => {
          e.preventDefault();
          const formData = new FormData(e.currentTarget);
          console.log(formData.getAll("topics"));
        }}
      >
        <label htmlFor="createTopic">Topics</label>
        <input
          type="text"
          value={topicValue}
          onChange={(e) => setTopicValue(e.target.value)}
          name="createTopic"
        />
        <button
          type="button"
          onClick={() => {
            const newTopics = [...topics, topicValue];
            setTopics(newTopics);
            setTopicValue("");
          }}
        >
          Create
        </button>
        {topics.map((topic, index) => {
          return (
            <Fragment key={index}>
              <label>
                {topic}
                <input type="checkbox" name="topics" value={topic} />
              </label>
            </Fragment>
          );
        })}
        <button type="submit">Submit</button>
      </form>
    </>
  );
}
