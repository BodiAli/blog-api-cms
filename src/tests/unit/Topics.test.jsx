import { screen, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Topics from "../../components/Topics/Topics";

describe("Topics component", () => {
  test("Should render a label, input, and button to create topics", () => {
    const { container } = render(<Topics />);

    expect(container).toMatchSnapshot();
  });

  test("Should create topic when button is clicked and topics is not empty", async () => {
    const user = userEvent.setup();

    render(<Topics />);

    const topicInput = screen.getByLabelText("Topics");
    const createButton = screen.getByRole("button", { name: "Create" });

    await user.type(topicInput, "Topic 1");
    await user.click(createButton);

    await user.type(topicInput, "Topic 2");
    await user.click(createButton);

    const topic1 = screen.getByLabelText("Topic 1");
    const topic2 = screen.getByLabelText("Topic 2");

    expect(topic1).toBeInTheDocument();
    expect(topic2).toBeInTheDocument();
  });
});
