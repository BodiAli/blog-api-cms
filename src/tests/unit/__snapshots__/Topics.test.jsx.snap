// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Topics component > Should render a label, input, and button to create topics 1`] = `
<div>
  <div>
    <div>
      <label>
        Topics
        <input
          placeholder="Topic name"
          type="text"
          value=""
        />
      </label>
      <button
        type="button"
      >
        Create
      </button>
    </div>
    <div
      style="display: none;"
    />
  </div>
</div>
`;
